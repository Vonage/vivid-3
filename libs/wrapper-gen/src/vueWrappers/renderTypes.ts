import { ComponentDef } from '../common/ComponentDef';
import { wrappedComponentName } from './name';

export default function renderTypes(components: ComponentDef[]) {
	const componentNames = components.map(wrappedComponentName);

	const typeImports = componentNames
		.flatMap((component) => [
			`import type { ${component} } from './components';`,
			`import type { ${component}Props, ${component}Events } from './components/${component}.types';`,
		])
		.join('\n');

	const propTypes = componentNames
		.map((component) => `	${component}: ${component}Props;`)
		.join('\n');

	const eventTypes = componentNames
		.map((component) => `	${component}: ${component}Events;`)
		.join('\n');

	const slotTypes = componentNames
		.map((component) => `	${component}: ComponentSlots<typeof ${component}>;`)
		.join('\n');

	return `
${typeImports}
import type { ComponentSlots } from '../utils/types';

export type VProps = {
${propTypes}
};

export type VEvents = {
${eventTypes}
};

export type VSlots = {
${slotTypes}
};
`;
}
