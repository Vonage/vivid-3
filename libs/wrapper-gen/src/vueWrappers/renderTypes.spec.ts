import {
	exampleComponent,
	minimalComponent,
} from '../__fixtures__/componentDefs';
import renderTypes from './renderTypes';

describe('renderTypes', () => {
	it('should render a files exporting all component types', () => {
		expect(renderTypes([minimalComponent, exampleComponent])).toMatchSnapshot();
	});
});
