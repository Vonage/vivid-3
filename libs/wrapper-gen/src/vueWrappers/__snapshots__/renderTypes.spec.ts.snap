// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renderTypes > should render a files exporting all component types 1`] = `
"
import type { VExampleComponent } from './components';
import type { VExampleComponentProps, VExampleComponentEvents } from './components/VExampleComponent.types';
import type { VExampleComponent } from './components';
import type { VExampleComponentProps, VExampleComponentEvents } from './components/VExampleComponent.types';
import type { ComponentSlots } from '../utils/types';

export type VProps = {
	VExampleComponent: VExampleComponentProps;
	VExampleComponent: VExampleComponentProps;
};

export type VEvents = {
	VExampleComponent: VExampleComponentEvents;
	VExampleComponent: VExampleComponentEvents;
};

export type VSlots = {
	VExampleComponent: ComponentSlots<typeof VExampleComponent>;
	VExampleComponent: ComponentSlots<typeof VExampleComponent>;
};
"
`;
